### PR TITLE
[Scope Granularity: Part 1 of 5]: Add service module, util and model support

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR is part of https://github.com/hypothesis/product-backlog/issues/908 and based on https://github.com/hypothesis/h/pull/5577

This piece of the feature _adds_ functionality but does not integrate any of it—it's standalone. These commits:

* Add a new `GroupScope` service
* Add a computed `scope` property to the `GroupScope` model
* Adds some additional scope-related functions to the `group_scope` util module

Feedback from https://github.com/hypothesis/h/pull/5577 has been integrated WRT using `startswith`/string matching versus true regexes. It is indeed simpler and will work nicely when we conflate to a single `scope` column.

### CODE REVIEW

**Comment by lyzadanger** [comment]
Marking WIP as the entire set of PRs for this features need to be merged and squashed in a particular order.

File: h/util/group_scope.py [hunk 1]
```diff
@@ -18,25 +18,73 @@ def match(uri, scopes):
...
+    :arg uri: URI string in question
+    :arg scopes: List of URIs that define scope
+    :type scopes: list(str)
+    :rtype: bool
+    """
+    scope_matches = [scope for scope in scopes if uri.startswith(scope)]
+
+    if len(scope_matches):
+        return True
+    return False
```
Comment (seanh) [inline h/util/group_scope.py]: I think the idiomatic way to do this in Python is by using the `any` builtin and a list or generator comprehension: `return any((uri.startswith(scope) for scope in scopes))`, which also avoids calling `startswith()` for every scope if the first one matches

File: h/util/group_scope.py [hunk 2]
```diff
@@ -28,6 +30,30 @@ def uri_scope(uri):
...
+    # string for path.
+    origin = _parse_origin(uri)
+    path = _parse_path(uri) or None
+    return (origin, path)
+
+
+def _parse_path(uri):
+    """Return the path component of a URI string"""
+    if uri is None:
+        return None
```
Comment (seanh) [inline h/util/group_scope.py]: Is `uri` ever `None` here? The tests still pass if I replace the function with `return urlparse.urlsplit(uri)[2]`, and I can't see from the code where `None` might be passed.

File: h/services/group_scope.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+from h.models import GroupScope
+from h.util import group_scope as scope_util
```
Comment (seanh) [inline h/services/group_scope.py]: Renaming things like this makes the code hard to grep

File: h/services/group_scope.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        """Return all GroupScope records that match on ``origin``
+
+        :arg uri: URI whose origin should be matched for
+        :type uri: str
+        :rtype: list(:class:`~h.models.group_scope.GroupScope`)
+        """
+        # Retrieve all GroupScope records in the DB that have an `origin` component
+        # that matches the given uri's origin. This will give us a set of
+        # results whose scopes may match the `document_uri`
+        # Refactor into scopes_for_origin
```
Comment (seanh) [inline h/services/group_scope.py]:
```suggestion
```

File: h/services/group_scope.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+from h.models import GroupScope
+from h.util import group_scope as scope_util
+
+
+class GroupScopeService(object):
+    def __init__(self, session):
+        self._session = session
+
+    def fetch_by_origin(self, uri):
```
Comment (seanh) [inline h/services/group_scope.py]: Is `fetch_by_origin()` intended to be a public function or is it only used by `fetch_by_scope()` below?

File: h/services/group_scope.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        """Return all GroupScope records that match on ``origin``
```
Comment (seanh) [inline h/services/group_scope.py]:
```suggestion
        """Return all GroupScope records that match on ``uri``'s origin
```

File: h/services/group_scope.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+        :arg uri: URI to find matching scopes for
+        :type uri: str
+        :rtype: list(:class:`~h.models.group_scope.GroupScope`)
+        """
+        origin_scopes = self.fetch_by_origin(uri)
+        return [
+            scope
+            for scope in origin_scopes
+            if scope_util.uri_in_scope(uri, [scope.scope])
```
Comment (seanh) [inline h/services/group_scope.py]: This seems to be the only place where `scope_util.uri_in_scope()` gets called and we're passing it a list of one scope. Could this be `if uri.startswith(scope.scope)`, and then `uri_in_scope()` wouldn't be needed?

File: tests/h/services/group_scope_test.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+class TestGroupScopeFactory(object):
+    def test_it_returns_group_scope_service_instance(self, pyramid_request):
+        svc = group_scope_factory(None, pyramid_request)
+
+        assert isinstance(svc, GroupScopeService)
+
+
+@pytest.fixture
+def svc(db_session, pyramid_request):
+    pyramid_request.db = db_session
```
Comment (seanh) [inline tests/h/services/group_scope_test.py]:
```suggestion
```

`pyramid_request.db` and `db_session` are already the same object

File: tests/h/services/group_scope_test.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            assert matches == []
+
+
+class TestFetchByScope(object):
+    def test_it_proxies_to_fetch_by_origin(self, svc, document_uri):
+        svc.fetch_by_origin = mock.Mock(return_value=[])
+
+        svc.fetch_by_scope(document_uri)
+
+        svc.fetch_by_origin.assert_called_once_with(document_uri)
```
Comment (seanh) [inline tests/h/services/group_scope_test.py]: I think this is testing an internal implementation detail, and with a dodgy mock

**Comment by lyzadanger** [comment]
@seanh 

> The tests seem to be failing for me?

There is one flaky test on this branch that is rectified in part 2 of this feature. It fails about once out of every 12 times I run the tests. Given that this is fixed in part 2 and I am planning on squashing all of these together, I didn't want to spend the time to debug it on this branch. Let me know if that doesn't fly.

**Pull request closed without merging**